### PR TITLE
feat: add support for plugin index

### PR DESCRIPTION
**Pull request by knqyf263**
## Description
This PR adds support for the plugin index. See https://github.com/aquasecurity/trivy/issues/6641 for more details.

## Usage

1. Download the plugin list:

```bash
$ trivy plugin update
```

2. Discover plugins available on the index:

```bash
$ trivy plugin search
NAME                 DESCRIPTION                                                  MAINTAINER           OUTPUT
aqua                 A plugin for integration with Aqua Security SaaS platform    aquasecurity
kubectl              A plugin scanning the images of a kubernetes resource        aquasecurity
referrer             A plugin for OCI referrers                                   aquasecurity           ✓
[...]
```

3. Choose a plugin from the list and install it:

```bash
$ trivy plugin install referrer
```

4. Use the installed plugin:

```bash
$ trivy referrer --help
```

5. Keep your plugins up-to-date:

```bash
$ trivy plugin upgrade
```

6. Uninstall a plugin you no longer use:

```bash
trivy plugin uninstall referrer
``` 

## Related issues
- Close https://github.com/aquasecurity/trivy/issues/6641

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

**Comment by knqyf263** [comment]
@DmitriyLewen I'm still writing a developer guide, but the implementation has been completed. I would appreciate your quick feedback.

**Review by DmitriyLewen — COMMENTED** [review]
Left some comments.

Also prefix log is broken:
```bash
➜  ./trivy plugin install aqua
2024/05/13 12:40:43 INFO Installing the plugin... prefix="[plugin] " src=github.com/aquasecurity/trivy-plugin-aqua
```

**Comment by DmitriyLewen** [comment]
1 more thought:
perhaps we want to add information that the plugin was installed successfully?

Currently:
```
➜  trivy git:(feat/plugin_index) ✗ ./trivy plugin install kubectl
2024/05/13 12:57:56 INFO Installing the plugin... prefix="[plugin] " src=github.com/aquasecurity/trivy-plugin-kubectl

```

File: pkg/commands/app.go [hunk 1]
```diff
@@ -789,22 +790,45 @@ func NewPluginCommand() *cobra.Command {
...
 			RunE: func(cmd *cobra.Command, args []string) error {
-				return plugin.RunWithURL(cmd.Context(), args[0], plugin.RunOptions{Args: args[1:]})
+				return plugin.RunWithURL(cmd.Context(), args[0], plugin.Options{Args: args[1:]})
 			},
 		},
 		&cobra.Command{
-			Use:                   "update PLUGIN_NAME",
-			Short:                 "Update an existing plugin",
+			Use:                   "update",
+			Short:                 "Update the local copy of the plugin index",
```
[the remark this answers is not in the document]
Comment (knqyf263) [reply pkg/commands/app.go:798]: It hsould be ok as people are familiar with `apt-get update`.

[the remark this answers is not in the document]
Comment (knqyf263) [reply pkg/commands/app.go:798]: >It’s also worth pointing out that we are only updating the cachedir/trivy/plugin/index.yaml file.

This is mentioned.

> Update the local copy of the plugin index

[the remark this answers is not in the document]
Comment (knqyf263) [reply pkg/commands/app.go:798]: Not only `apt`, but also [brew](https://docs.brew.sh/Manpage#update-options), [krew](https://krew.sigs.k8s.io/docs/user-guide/quickstart/), and many other tools. But we can still update it if it confuses users.

[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply pkg/commands/app.go:798]: my concern is that we will change the logic for `update`.
these changes may confuse users who currently use plugins.

**Comment by knqyf263** [comment]
>Also prefix log is broken:

Thanks. It worked at the beginning, and got broken after refactoring. Fixed in https://github.com/aquasecurity/trivy/pull/6674/commits/9dbecdfa41ecfc5868860d291c38664158cc80ba

File: pkg/plugin/manager.go [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+}
+func RunWithURL(ctx context.Context, name string, opts Options) error {
+	return defaultManager.RunWithURL(ctx, name, opts)
+}
+func Upgrade(ctx context.Context, names []string) error { return defaultManager.Upgrade(ctx, names) }
+func Uninstall(ctx context.Context, name string) error  { return defaultManager.Uninstall(ctx, name) }
+func Information(name string) error                     { return defaultManager.Information(name) }
+func List(ctx context.Context) error                    { return defaultManager.List(ctx) }
+func Update(ctx context.Context) error                  { return defaultManager.Update(ctx) }
+func Search(ctx context.Context, args []string) error   { return defaultManager.Search(ctx, args) }
```
[the remark this answers is not in the document]
Comment (knqyf263) [reply pkg/plugin/manager.go]: No, I just passed through the CLI arguments, but it looks better to pass one argument. Fixed in https://github.com/aquasecurity/trivy/pull/6674/commits/d2f2e963c3bb63993da8cfb922fb1a5c9cf1261b.

File: pkg/commands/app.go [hunk 3]
```diff
@@ -789,22 +790,45 @@ func NewPluginCommand() *cobra.Command {
...
+			RunE: func(cmd *cobra.Command, args []string) error {
+				return plugin.Search(cmd.Context(), args)
+			},
+		},
+		&cobra.Command{
+			Use:                   "upgrade [PLUGIN_NAMES]",
+			Short:                 "Upgrade installed plugins to newer versions",
+			SilenceErrors:         true,
+			DisableFlagsInUseLine: true,
+			Args:                  cobra.MaximumNArgs(1),
```
[the remark this answers is not in the document]
Comment (knqyf263) [reply pkg/commands/app.go]: You're right. Thanks. Deleted https://github.com/aquasecurity/trivy/pull/6674/commits/2fcd08ba92d89ad24f6ced49adc8c32f20fff731

File: docs/docs/plugin/user-guide.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+```bash
+$ trivy plugin search
+NAME                 DESCRIPTION                                                  MAINTAINER           OUTPUT
+aqua                 A plugin for integration with Aqua Security SaaS platform    aquasecurity
+kubectl              A plugin scanning the images of a kubernetes resource        aquasecurity
+referrer             A plugin for OCI referrers                                   aquasecurity           ✓
+```
+
+You can specify search keywords as arguments:
```
[the remark this answers is not in the document]
Comment (knqyf263) [reply docs/docs/plugin/user-guide.md:23]: Added https://github.com/aquasecurity/trivy/pull/6674/commits/39ce40e6b5b6eee60bda488f10585a8aa06fba12

**Comment by knqyf263** [comment]
>perhaps we want to add information that the plugin was installed successfully?

Done
https://github.com/aquasecurity/trivy/pull/6674/commits/466b9f5ea59b8fdeffe4ab2c2cabb0f4f22aaf79

**Comment by knqyf263** [comment]
@DmitriyLewen I've addressed your comments and completed the developer guide. This PR is now ready for review.

File: pkg/plugin/manager.go [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+}
+
+func (m *Manager) upgrade(ctx context.Context, name string) error {
+	plugin, err := m.load(name)
+	if err != nil {
+		return xerrors.Errorf("plugin load error: %w", err)
+	}
+
+	logger := log.With("name", name)
+	logger.InfoContext(ctx, "Upgrading plugin...")
```
Comment (DmitriyLewen) [inline pkg/plugin/manager.go:222]:
```suggestion
	m.logger.InfoContext(ctx, "Upgrading plugin...", log.String("name", name))
```

File: docs/docs/references/configuration/cli/trivy_plugin_search.md
Comment (DmitriyLewen) [inline docs/docs/references/configuration/cli/trivy_plugin_search.md:1]: Add this page to [config file](https://github.com/aquasecurity/trivy/blob/150a77313e980cd63797a89a03afcbc97b285f38/mkdocs.yml#L158-L164) please.

File: docs/docs/references/configuration/cli/trivy_plugin_upgrade.md
Comment (DmitriyLewen) [inline docs/docs/references/configuration/cli/trivy_plugin_upgrade.md:1]: Add this page to [config file](https://github.com/aquasecurity/trivy/blob/150a77313e980cd63797a89a03afcbc97b285f38/mkdocs.yml#L158-L164) please.

File: docs/docs/plugin/index.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+```bash
+$ trivy plugin upgrade
+```
+
+6. Uninstall a plugin you no longer use:
+
+```bash
+trivy plugin uninstall referrer
+```
```
Comment (DmitriyLewen) [inline docs/docs/plugin/index.md:63]: numbering is broken.
```suggestion
    ```bash
    $ trivy plugin update
    ```

2. Discover Trivy plugins available on the plugin index:

    ```bash
    $ trivy plugin search
    NAME                 DESCRIPTION                                                  MAINTAINER           OUTPUT
    aqua                 A plugin for integration with Aqua Security SaaS platform    aquasecurity
    kubectl              A plugin scanning the images of a kubernetes resource        aquasecurity
    referrer             A plugin for OCI referrers                                   aquasecurity           ✓
    [...]
    ```

3. Choose a plugin from the list and install it:

    ```bash
    $ trivy plugin install referrer
    ```

4. Use the installed plugin:

    ```bash
    $ trivy referrer --help
    ```

5. Keep your plugins up-to-date:

    ```bash
    $ trivy plugin upgrade
    ```

6. Uninstall a plugin you no longer use:

    ```bash
    trivy plugin uninstall referrer
    ``` 
```

File: docs/docs/plugin/user-guide.md [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+- XDG_DATA_HOME if set and .trivy/plugins exists within the XDG_DATA_HOME dir
+- ~/.trivy/plugins
+
+Furthermore, it is possible to download plugins that are not registered in the index by specifying the URL directly or by specifying the file path.
+
+```bash
+$ trivy plugin install github.com/aquasecurity/trivy-plugin-kubectl
+```
+`
```
Comment (DmitriyLewen) [inline docs/docs/plugin/user-guide.md:55]:
```suggestion
```

File: docs/docs/plugin/index.md (hunk 6 above)
Comment (knqyf263) [reply docs/docs/plugin/index.md]: Thanks!

File: pkg/plugin/manager.go [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+
+// Uninstall installs the plugin
+func (m *Manager) Uninstall(ctx context.Context, name string) error {
+	pluginDir := filepath.Join(m.pluginRoot, name)
+	if !fsutils.DirExists(pluginDir) {
+		m.logger.ErrorContext(ctx, "No such plugin")
+		return nil
+	}
+	return os.RemoveAll(pluginDir)
+}
```
[the remark this answers is not in the document]
Comment (knqyf263) [reply pkg/plugin/manager.go]: Added https://github.com/aquasecurity/trivy/pull/6674/commits/2fe4f21ad27392e63b0115413a8d83484737a19d

File: pkg/commands/app.go
[the remark this answers is not in the document]
Comment (knqyf263) [reply pkg/commands/app.go:1]: Added https://github.com/aquasecurity/trivy/pull/6674/commits/1835268ff46527a693fb3635c2fb65067e5f4725

File: pkg/plugin/manager.go (hunk 5 above)
Comment (knqyf263) [reply pkg/plugin/manager.go]: Fixed https://github.com/aquasecurity/trivy/pull/6674/commits/60f4848f546f7e9de9ccd58f96762faebac4b8b2

File: docs/docs/references/configuration/cli/trivy_plugin_search.md
Comment (knqyf263) [reply docs/docs/references/configuration/cli/trivy_plugin_search.md:1]: Grouped plugin subcommands https://github.com/aquasecurity/trivy/pull/6674/commits/a240e5268c8b8040be13dd5624c9acdb8dbad2f1

File: docs/docs/references/configuration/cli/trivy_plugin_upgrade.md
Comment (knqyf263) [reply docs/docs/references/configuration/cli/trivy_plugin_upgrade.md:1]: Grouped plugin subcommands https://github.com/aquasecurity/trivy/pull/6674/commits/a240e5268c8b8040be13dd5624c9acdb8dbad2f1

**Review by DmitriyLewen — APPROVED** [review]
LGTM

**Pull request merged**